File: intern/515.find-largest-value-in-each-tree-row.js
/*
 * @lc app=leetcode id=515 lang=javascript
 *
 * [515] Find Largest Value in Each Tree Row
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var largestValues = function (root) {
    if (root === null) {
        return [];
    }
    const queue = [[root, 0]];
    const max = [];

    while (queue.length > 0) {
        const [node, layer] = queue.shift();
        max[layer] = Math.max(max[layer] ?? -Infinity, node.val);

        if (node.left !== null) {
            queue.push([node.left, layer + 1]);
        }
        if (node.right !== null) {
            queue.push([node.right, layer + 1]);
        }
    }

    return max;
};
// @lc code=end

File: Daily/515.find-largest-value-in-each-tree-row.cpp
/*
 * @lc app=leetcode id=515 lang=cpp
 *
 * [515] Find Largest Value in Each Tree Row
 */

#include <cmath>
#include <vector>

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : val(x), left(left), right(right) {}
};

// @lc code=start
class Solution {
 public:
  std::vector<int> largestValues(TreeNode* root) {
    std::vector<int> maxValuesInLevel;
    dfs(root, 0, maxValuesInLevel);
    return maxValuesInLevel;
  }

 private:
  static void dfs(TreeNode* root,
                  const int level,
                  std::vector<int>& /*out*/ maxValuesInLevel) {
    if (root == nullptr) {
      return;
    }
    if (maxValuesInLevel.size() == level) {
      maxValuesInLevel.push_back(root->val);
    } else {
      maxValuesInLevel[level] = std::max(maxValuesInLevel[level], root->val);
    }

    if (root->left) {
      dfs(root->left, level + 1, maxValuesInLevel);
    }
    if (root->right) {
      dfs(root->right, level + 1, maxValuesInLevel);
    }
  }
};
// @lc code=end
